src/server/entities/ShieldFragmentServerEntity.java

```diff
@@ -2,8 +2,44 @@
 
 import common.EntityType;
 
-public class ShieldFragmentServerEntity extends ServerEntity{
-    public ShieldFragmentServerEntity(float X, float Y) {
+class Placeholder{
+    String text;
+    int number;
+
+    public Placeholder(String text, int number){
+        this.text = text;
+        this.number = number;
+    }
+
+    public Placeholder(Placeholder placeholder){
+        this.text = placeholder.text;
+        this.number = placeholder.number;
+    }
+}
+
+public class ShieldFragmentServerEntity extends ServerEntity implements Cloneable{
+    public Placeholder placeholder;
+
+    public ShieldFragmentServerEntity(float X, float Y, String text, int number) {
         super(EntityType.SHIELD, X, Y, 10, 10, 0, 0);
+        placeholder = new Placeholder(text, number);
+    }
+
+    public ShieldFragmentServerEntity(ShieldFragmentServerEntity e){
+        super(e);
+        placeholder = new Placeholder(e.placeholder);
+    }
+
+    public ShieldFragmentServerEntity shallowCopy(){
+        try{
+            return (ShieldFragmentServerEntity) this.clone();
+        } catch(CloneNotSupportedException ex){
+            ex.printStackTrace();
+            return this;
+        }
+    }
+
+    public ShieldFragmentServerEntity deepCopy(){
+        return new ShieldFragmentServerEntity(this);
```

Comment:
Suggestion to make the Placeholder class implement the Cloneable interface too, so we can call clone on it whenever we make a deep copy. This means that inside `deepCopy()`, we'd call `shallowCopy()` to copy the ShieldFragmentServerEntity, and set its placeholder object reference to `placeholder.clone()`. Reference: [https://www.baeldung.com/java-deep-copy#2-cloneable-interface](url)